[Woo POS] Stop `GhostableCardView` rendering, and network requests when there are no new items

iamgabrielma opened the pull request:
## Description
This PR makes the GhostableCardView to only rely on `isLoading` state  in order to render, as well as marking the last page when no new unique items are fetched from the remote. This resolves two issues:
- The `GhostableCardView` will no longer render indefinitely when we reach the end of the item list
- We stop doing network requests unless we reload the item list, in which case `isLastPage` resets.

~~Caveat: I think there’s a compromise to be made with our current limitations on eligible products: if there’s a gap in the API response between eligible products we might trigger the “last page” too early (this is easily reproducible if we limit products per page to something like 5, and then there's 5+ non-eligible products in the request).~~

~~We can see this in the video below where the  GhostableCardView doesn't render for a second if there's a gap, but then we immediately fetch next page, it has items, so loads the rest normally.  This should not be a problem in release, as would imply that the merchant has a 100+ product gap that are not eligible between page fetches. This can be improved in the future as needed, but any advice or recommendation of how to handle it better is welcome.~~

Update: these have been resolved in the latest iteration

https://github.com/user-attachments/assets/4b6fb126-1037-49f9-98bd-bf0eef3ae62f



## Steps to reproduce
- With a POS eligible store, for simplicity modify `POSConstants.productsPerPage` to something like 5 or 10, and load POS
- Scrolling down should show the GhostableCardView while there are more items to fetch
- At the end of the list, the GhostableCardView should disappear
- Pull-to-refresh should reload the 1st page normally, scrolling down should show the GhostableCardView while there are more items to fetch, and stop at the end of the list.
- If we have a store with 1-2 products, the `GhostableCardView` will also stop as soon as the end of the page is triggered ( you can create a new jurassic ninja site with 1 simple product, or I can send access to a test site if needed 👍 )

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
Set isLastPage when no new items are fetched from remote

joshheald commented on WooCommerce/Classes/POS/Presentation/ItemListView.swift:
```diff
@@ -134,15 +134,15 @@ private extension ItemListView {
                 }
                 GhostItemCardView()
-                    .renderedIf(posModel.itemListState.isLoadingAfterInitialLoad)
+                    .renderedIf(posModel.itemListState.isLoadingAfterInitialLoad && !posModel.isLastPage)
```
This seems a bit off... 

Why do we need to check what page it's loading if the state is a loading state?

The `itemListState` shouldn't be any kind of loading if there's nothing left to load; it should just go directly to `loaded`.

iamgabrielma replied:
True that, the `isLastPage` property shouldn't influence here, and leave the View to just render when state is loading, shouldn't have to know about anything else.

joshheald commented on WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift:
```diff
@@ -73,6 +75,12 @@ extension PointOfSaleAggregateModel {
+            isLastPage = true
+        } else {
+            isLastPage = false
+        }
```
We should lobby for `X-WP-Total` and `X-WP-TotalPages` headers to be passed through the JP tunnel, instead of trying to figure this out for ourselves.

joshheald reviewed (changes requested):
I don't think this is the right approach – it still results in unlimited network requests, and can get the `isLastPage` decision wrong. 

We will keep our views and logic easier to understand and change if we avoid adding flags like `isLastPage`, or at least avoid using them in views – the itemListState can give everything we need here. It should be either `loaded` or `loading`, depending on whether we want to show the ghost cell or not. 

If we don't want to show it, but still do a network request, we can just leave it as `loaded` while that request continues.

I've done some exploration of it and it would work better if we determine whether we're at the end of the list in the service layer, since there it can throw a `pageOutOfRange` error if we get no products back in the API response. That avoids the client-side filtering having an effect.

We may still need some private state in the aggregate model, but it doesn't need to be published or even available to the view. It should just inform what we publish in the `itemListState` property. 

Give me a shout if you want to pair on it tomorrow, especially if any of this isn't clear. I've got my exploration on a branch but it's messy and I want to focus on the cart and other fixes today, rather than polishing it.

iamgabrielma committed:
Merge branch 'trunk' into task/infinite-scroll-derender-ghostable-view-when-last-page

# Conflicts:
#	WooCommerce/Classes/POS/Presentation/ItemListView.swift

iamgabrielma committed:
throw pageIsOutOfRange error to stop fetching

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/232">21.2</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

iamgabrielma committed:
make test target compile

iamgabrielma committed:
Do not throw out of range error for no eligible products

If there are no eligible products at all (page = 1 && products = 0), we would throw the out of range error, rather than returning empty.

iamgabrielma committed:
Add test cases

iamgabrielma committed:
cleanup loadNextItems

iamgabrielma committed:
Merge branch 'trunk' into task/infinite-scroll-derender-ghostable-view-when-last-page

# Conflicts:
#	WooCommerce/Classes/POS/Models/PointOfSaleAggregateModel.swift
#	WooCommerce/WooCommerceTests/POS/Models/PointOfSaleAggregateModelTests.swift

iamgabrielma committed:
revert loadNextItems() refactor

Behavior is the same, but some tests that expect `.empty` state fail with `.loaded([])` instead. Effectively these mean the same but let’s defer it outside of the scope of this PR.

iamgabrielma committed:
Resolve add missing tests after merge conflict

iamgabrielma committed:
lint

iamgabrielma commented:
Thanks for the review and the pairing @joshheald , this is ready for review when you have the chance. No logic has been changed since yesterday except for [this commit](https://github.com/woocommerce/woocommerce-ios/pull/14382/commits/668985b323f083b75856791ee818dd0f88035a49), which checks  we're not throwing the out of range error if we just happen to have no eligible products at all. 

https://github.com/user-attachments/assets/cf2b1579-02ce-47d3-bd62-ae1e5f3f4d82

When cleaning up the `loadNextItems()` method here https://github.com/woocommerce/woocommerce-ios/pull/14382/commits/8956c01a42484791edcd27bc1c6876c7c2fad267 on my testing I could see the behaviour was exactly the same, however some tests started to fail because expected `.empty` but received `.loaded([])` instead, so I [reverted](https://github.com/woocommerce/woocommerce-ios/pull/14382/commits/4c0a76db63d8ac75b6de58c8f8a396edec57e46a) it for now, this can be done separately once the design/arch changes are completed.

iamgabrielma committed:
Add missing expectation to test

We don’t do anything specific with this error in the case that the page is out of range, but we still use the same associated type when there’s an error on loading further products.

iamgabrielma committed:
restore rendering GhostItemCardView behind flag

joshheald commented on Yosemite/Yosemite/PointOfSale/POSProductProvider.swift:
```diff
@@ -32,25 +38,20 @@ public final class POSProductProvider: POSItemProvider {
-        } catch {
-            DDLogError("Failed to retrieve products. Error: \(error)")
-            throw POSProductProviderError.requestFailed
+        if pageNumber != 1 && products.count == 0 {
```
I could argue both ways on this one (we could leave it to the view layer to show an `outOfRange` error with nothing previously loaded as the empty state) but this looks fine to me. 👍

joshheald committed:
Tidy up last page loaded logic

joshheald reviewed (approved):
@iamgabrielma Thanks for the ping, sorry I didn't get to it yesterday.

Tests well. I tidied up the logic a bit:
- renamed and inverted the flag from `pageIsOutOfRange` to `mightHaveMorePages`.
- Moved the "error" handling and flag management to the `load(pageNumber:) function so it happens even on initial load.
- Extracted the `itemListState` determination to a function which we call from `load(pageNumber:)`, instead of doing it in `fetch` and `loadNextPage`.

I didn't write an extra test for the `initialPageLoad` cases here; perhaps that would be worthwhile to do next week?

Also I found a "fun" edge case...
If the first page of results doesn't have any products that pass the filter, we show the empty state... but that means we never load page 2 even if it might have something.

It might be solvable, because I don't _think_ the item provider will tell the aggregate model that there's no more pages in this case... so we could try triggering subsequent page loads until we get the `pageOutOfRange` error (even though there's no view to trigger those for us) and only show the empty state when we get `pageOutOfRange` and have an empty `allItems` array.

I'm going to merge this for now, but perhaps you could ticket and prioritise these things?

iamgabrielma commented:
Thanks tidying up things and merging, appreciated! 

> I didn't write an extra test for the initialPageLoad cases here; perhaps that would be worthwhile to do next week?

Definitely. Logged here: https://github.com/woocommerce/woocommerce-ios/issues/14441

> I found a "fun" edge case...If the first page of results doesn't have any products that pass the filter, we show the empty state... but that means we never load page 2 even if it might have something.

Yes, I think we could try triggering a couple of subsequent pages automatically to address this edge case, or perhaps offer a "retry" button that does this. It would be a better experience than getting stuck without knowing why. Logged: https://github.com/woocommerce/woocommerce-ios/issues/14442
